vscode/src/editor/utils/folding-ranges.ts

```diff
@@ -0,0 +1,228 @@
+import * as vscode from 'vscode'
+
+import { getFoldingRanges, getSymbols } from '.'
+
+/**
+ * Gets the folding range containing the target position.
+ *
+ * NOTE: Use getSmartSelection from utils/index.ts instead
+ *
+ * @param uri - The URI of the document to get folding ranges for
+ * @param target - The target position number
+ * @returns The folding range containing the target position, or undefined if none found
+ */
+export async function getTargetFoldingRange(uri: vscode.Uri, target: number): Promise<vscode.Selection | undefined> {
+    // Check if symbols is available in the doc by its file uri and look for language id
+    const doc = await vscode.workspace.openTextDocument(uri)
+    // Remove text-based languages that don't support symbols or folding ranges
+    const textBaseLangIds = ['markdown', 'plaintext', 'json', 'sql']
+    const isSupported = !textBaseLangIds.includes(doc.languageId)
+
+    // Get the ranges of all classes and folding ranges in parallel
+    const [ranges, classes] = await Promise.all([
+        getFoldingRanges(uri).then(r => r?.filter(r => !r.kind)),
+        isSupported
+            ? getSymbols(uri)
+                  .then(r => r?.filter(s => s.kind === vscode.SymbolKind.Class))
+                  .then(s => s?.map(symbol => symbol.location.range))
+            : [],
+    ])
+
+    // If lanaguage id not available, find class object ranges heuristically
+    const classRanges = isSupported ? classes : await getOuterClassFoldingRanges(ranges, uri)
```

Comment:
```suggestion
    const classRanges = doc.languageId ? classes : await getOuterClassFoldingRanges(ranges, uri)
```

isSupported is the check for textBasedLanguages but you wanna run `getOuterClassFoldingRanges` for languages that don't support langserver and you want to do that when the language id is not available so then you use heuristics to find class ranges.

In the case of Json, plaintext, sql, markdown the whole concept of classes doesn't make sense so you get empty array of classes(which is what you want)

Reply:
doc.languageId returns 'plaintext' for code files that do not have lang server support, which I am trying to capture here. I will add a comment to reflect this, thanks!